src/core/servers.c

```diff
@@ -276,48 +272,37 @@ static void server_real_connect(SERVER_REC *server, IPADDR *ip,
 	}
 }
 
-static void server_connect_callback_readpipe(SERVER_REC *server)
+static void server_connect_callback_readpipe(RESOLVED_IP_REC *iprec, SERVER_REC *server)
 {
-	RESOLVED_IP_REC iprec;
-        IPADDR *ip;
+	IPADDR *ip;
 	const char *errormsg;
 
-	g_source_remove(server->connect_tag);
-	server->connect_tag = -1;
-
-	net_gethostbyname_return(server->connect_pipe[0], &iprec);
-
-	g_io_channel_shutdown(server->connect_pipe[0], TRUE, NULL);
-	g_io_channel_unref(server->connect_pipe[0]);
-	g_io_channel_shutdown(server->connect_pipe[1], TRUE, NULL);
-	g_io_channel_unref(server->connect_pipe[1]);
-
-	server->connect_pipe[0] = NULL;
-	server->connect_pipe[1] = NULL;
+	server->connect_cancellable = NULL;
 
 	/* figure out if we should use IPv4 or v6 address */
-	if (iprec.error != 0) {
-                /* error */
+	if (iprec->error != NULL) {
```

Comment:
This comment here is a symptom of the problem. irssi code should not need to figure this out: the resolver gives an order to try and we should try it.